File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Subsystems/TurretSubsystem.java
package org.firstinspires.ftc.teamcode.Subsystems;

import com.pedropathing.geometry.Pose;

import dev.nextftc.control.ControlSystem;
import dev.nextftc.control.KineticState;
import dev.nextftc.core.subsystems.Subsystem;
import dev.nextftc.hardware.impl.MotorEx;

public class TurretSubsystem implements Subsystem {
    public static TurretSubsystem INSTANCE = new TurretSubsystem();
    private TurretSubsystem() { }

    public MotorEx turretMotor = new MotorEx("tur");
    public ControlSystem turretControl = ControlSystem.builder()
            .posPid(0.012)
            .build()
    ;

    public static double LOOP_LIMIT = 0;
    public static double LOOP_INDEX = 0;
    public static double LAST_RECORDED_TICKS = 0;

    public void calculateHeading(Pose currentPos) {
        double x = currentPos.getX();
        double y = currentPos.getY();
        double heading = currentPos.getHeading();
        double angle = Math.atan((-48-y)/(72-x));
        double ticks = Math.round(((angle-heading)/(2*Math.PI)) * 384.5 * (100/24));
        if (Math.abs(ticks-LAST_RECORDED_TICKS) >= 3) {
            turretControl.setGoal(new KineticState(LAST_RECORDED_TICKS));
            LAST_RECORDED_TICKS = ticks;
        }
    }

    @Override
    public void periodic() {
        if (LOOP_INDEX == LOOP_LIMIT) {
            turretMotor.setPower(turretControl.calculate(turretMotor.getState()));
            LOOP_INDEX = 0;
        } else {
            LOOP_INDEX ++;
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Systems/Subsystems/TurretSubsystem.kt
package org.firstinspires.ftc.teamcode.Systems.Subsystems

import com.pedropathing.geometry.Pose
import dev.nextftc.control.KineticState
import dev.nextftc.control.builder.controlSystem
import dev.nextftc.core.subsystems.Subsystem
import dev.nextftc.core.units.abs
import dev.nextftc.hardware.impl.MotorEx
import kotlin.math.*;

object TurretSubsystem: Subsystem {
    private val turretMotor = MotorEx("tur")
    private val turretControl = controlSystem {
        posPid(0.03, 0.0, 0.0)
    }
    var LAST_RECORDED_TICKS = 0.0;
    var LOOP_INDEX = 0
    val LOOP_LIMIT = 20

    fun setTurretHeading(currentPose: Pose) {
        var x = currentPose.x
        var y = currentPose.y
        var heading = currentPose.heading
        var angle = atan((-144 - y) / (144 - x))
        var ticks = round((angle - heading) / (2 * PI) * 384.5 * (100 / 24))
        if (abs(ticks - LAST_RECORDED_TICKS) >= 5) {
            turretControl.goal = KineticState(ticks)
            LAST_RECORDED_TICKS = ticks
        }
    }

    override fun periodic() {
        if (LOOP_INDEX == LOOP_LIMIT) {
            LOOP_INDEX = 0
            turretMotor.power = turretControl.calculate(turretMotor.state)
        } else {
            LOOP_INDEX++;
        }
    }
}